File: parseBetpawa.js
// parseBetpawa.js -- portage JS de parse_betpawa.py

function lignesNonVides(texte) {
  return texte.split("\n").map((l) => l.trim()).filter((l) => l.length > 0);
}

function versNombre(s) {
  const n = parseFloat(s.replace(",", "."));
  return Number.isNaN(n) ? null : n;
}

function lirePaires(lignes, i, n) {
  const resultat = {};
  let idx = i;
  for (let k = 0; k < n; k++) {
    if (idx + 1 >= lignes.length) break;
    const valeur = versNombre(lignes[idx + 1]);
    if (valeur === null) break;
    resultat[lignes[idx]] = valeur;
    idx += 2;
  }
  return { paires: resultat, index: idx };
}

function lirePairesJusquaRupture(lignes, i, motifLabel) {
  const resultat = {};
  let idx = i;
  while (idx + 1 < lignes.length && motifLabel.test(lignes[idx])) {
    const valeur = versNombre(lignes[idx + 1]);
    if (valeur === null) break;
    resultat[lignes[idx]] = valeur;
    idx += 2;
  }
  return { paires: resultat, index: idx };
}

function parseBetpawa(texte, nomDomicile, nomExterieur) {
  const lignes = lignesNonVides(texte);
  const cotes = {};
  let i = 0;
  const REGEX_OU = /^(Plus|Moins) de (\d+\.5)$/;
  const REGEX_OU_RUPTURE = /^(Plus|Moins) de \d+\.5$/;

  while (i < lignes.length) {
    const titre = lignes[i];

    if (titre === "1X2 | Fin de Match") {
      const r = lirePaires(lignes, i + 1, 3); i = r.index;
      if (Object.keys(r.paires).length === 3) cotes["1x2"] = { "1": r.paires["1"], "N": r.paires["X"], "2": r.paires["2"] };
      continue;
    }

    if (titre === "Double Chance | Fin de Match") {
      const r = lirePaires(lignes, i + 1, 3); i = r.index;
      if (Object.keys(r.paires).length === 3) cotes["double_chance"] = { "1N": r.paires["1X"], "N2": r.paires["X2"], "12": r.paires["12"] };
      continue;
    }

    if (titre === "Les Deux Équipes Marquent | Fin de Match") {
      const r = lirePaires(lignes, i + 1, 2); i = r.index;
      if (Object.keys(r.paires).length === 2) cotes["btts"] = { "Oui": r.paires["Oui"], "Non": r.paires["Non"] };
      continue;
    }

    if (titre === "Plus de/Moins de | Fin de Match") {
      const r = lirePairesJusquaRupture(lignes, i + 1, REGEX_OU_RUPTURE); i = r.index;
      for (const [label, valeur] of Object.entries(r.paires)) {
        const m = label.match(REGEX_OU);
        cotes[`over_under_${m[2]}`] = cotes[`over_under_${m[2]}`] || {};
        cotes[`over_under_${m[2]}`][m[1] === "Plus" ? "plus" : "moins"] = valeur;
      }
      continue;
    }

    const mEquipe = titre.match(/^Plus de\/Moins de \| (.+) \| Fin de [Mm]atch$/);
    if (mEquipe) {
      const r = lirePairesJusquaRupture(lignes, i + 1, REGEX_OU_RUPTURE); i = r.index;
      const prefixe = mEquipe[1] === nomDomicile ? "over_under_domicile" : (mEquipe[1] === nomExterieur ? "over_under_exterieur" : null);
      if (prefixe) for (const [label, valeur] of Object.entries(r.paires)) {
        const m = label.match(REGEX_OU);
        cotes[`${prefixe}_${m[2]}`] = cotes[`${prefixe}_${m[2]}`] || {};
        cotes[`${prefixe}_${m[2]}`][m[1] === "Plus" ? "plus" : "moins"] = valeur;
      }
      continue;
    }

    if (titre === "Handicap À 3 Choix | Fin de Match") {
      i += 1;
      while (i < lignes.length && ["1", "X", "2"].includes(lignes[i])) i += 1;
      while (i + 5 < lignes.length) {
        const l1 = lignes[i], l2 = lignes[i + 2], l3 = lignes[i + 4];
        if (!/^Domicile\s+-\d+$/i.test(l1) || !/^Nul\s+-\d+$/i.test(l2) || !/^Extérieur\s+\+\d+$/i.test(l3)) break;
        const ligne = parseInt(l1.match(/-\d+$/)[0].slice(1), 10);
        const vd = versNombre(lignes[i + 1]), vn = versNombre(lignes[i + 3]), ve = versNombre(lignes[i + 5]);
        if (vd === null || vn === null || ve === null) break;
        cotes[`handicap_3choix_${ligne}`] = { domicile: vd, nul: vn, exterieur: ve };
        i += 6;
      }
      continue;
    }

    if (titre === "Impair/Pair | Fin de Match") {
      const r = lirePaires(lignes, i + 1, 2); i = r.index;
      if (Object.keys(r.paires).length === 2) cotes["pair_impair"] = { pair: r.paires["Pair"], impair: r.paires["Impair"] };
      continue;
    }

    i += 1;
  }
  return cotes;
}

if (typeof module !== "undefined") module.exports = { parseBetpawa };

File: parse_betpawa.py
"""
parse_betpawa.py -- convertit un copier-coller BRUT de la page marchés
Betpawa en cotes_manuelles.

Le périmètre handicap retenu ici est exclusivement :
« Handicap À 3 Choix | Fin de Match ».
Les autres formes de handicap sont ignorées.
"""
import json
import re
import sys


def _lignes_non_vides(texte):
    return [l.strip() for l in texte.splitlines() if l.strip()]


def _paires(lignes, i, n):
    resultat = {}
    for _ in range(n):
        if i + 1 >= len(lignes):
            break
        label, valeur = lignes[i], lignes[i + 1]
        try:
            resultat[label] = float(valeur.replace(",", "."))
        except ValueError:
            break
        i += 2
    return resultat, i


def _paires_jusqua_rupture(lignes, i, motif_label):
    resultat = {}
    while i + 1 < len(lignes) and re.match(motif_label, lignes[i]):
        label, valeur = lignes[i], lignes[i + 1]
        try:
            resultat[label] = float(valeur.replace(",", "."))
        except ValueError:
            break
        i += 2
    return resultat, i


def parse_betpawa(texte, nom_domicile, nom_exterieur):
    lignes = _lignes_non_vides(texte)
    cotes = {}
    i = 0
    while i < len(lignes):
        titre = lignes[i]

        if titre == "1X2 | Fin de Match":
            paires, i = _paires(lignes, i + 1, 3)
            if len(paires) == 3:
                cotes["1x2"] = {"1": paires.get("1"), "N": paires.get("X"), "2": paires.get("2")}
            continue

        if titre == "Double Chance | Fin de Match":
            paires, i = _paires(lignes, i + 1, 3)
            if len(paires) == 3:
                cotes["double_chance"] = {"1N": paires.get("1X"), "N2": paires.get("X2"), "12": paires.get("12")}
            continue

        if titre == "Les Deux Équipes Marquent | Fin de Match":
            paires, i = _paires(lignes, i + 1, 2)
            if len(paires) == 2:
                cotes["btts"] = {"Oui": paires.get("Oui"), "Non": paires.get("Non")}
            continue

        if titre == "Plus de/Moins de | Fin de Match":
            paires, i = _paires_jusqua_rupture(lignes, i + 1, r"^(Plus|Moins) de \d+\.5$")
            for label, valeur in paires.items():
                m = re.match(r"^(Plus|Moins) de (\d+\.5)$", label)
                cle_sel = "plus" if m.group(1) == "Plus" else "moins"
                cotes.setdefault(f"over_under_{m.group(2)}", {})[cle_sel] = valeur
            continue

        m_equipe = re.match(r"^Plus de/Moins de \| (.+) \| Fin de [Mm]atch$", titre)
        if m_equipe:
            nom = m_equipe.group(1)
            prefixe = "over_under_domicile" if nom == nom_domicile else ("over_under_exterieur" if nom == nom_exterieur else None)
            paires, i = _paires_jusqua_rupture(lignes, i + 1, r"^(Plus|Moins) de \d+\.5$")
            if prefixe:
                for label, valeur in paires.items():
                    m = re.match(r"^(Plus|Moins) de (\d+\.5)$", label)
                    cotes.setdefault(f"{prefixe}_{m.group(2)}", {})["plus" if m.group(1) == "Plus" else "moins"] = valeur
            continue

        if titre == "Handicap À 3 Choix | Fin de Match":
            i += 1
            # En-têtes Betpawa : 1 / X / 2.
            while i < len(lignes) and lignes[i] in {"1", "X", "2"}:
                i += 1
            # Chaque ligne est : Domicile -N / Nul -N / Extérieur +N, puis 3 cotes.
            while i + 5 < len(lignes):
                labels = lignes[i], lignes[i + 2], lignes[i + 4]
                if not (
                    re.match(r"^Domicile\s+-\d+$", labels[0], re.I)
                    and re.match(r"^Nul\s+-\d+$", labels[1], re.I)
                    and re.match(r"^Extérieur\s+\+\d+$", labels[2], re.I)
                ):
                    break
                try:
                    ligne = float(re.search(r"-\d+$", labels[0]).group()[1:])
                    cotes[f"handicap_3choix_{int(ligne)}"] = {
                        "domicile": float(lignes[i + 1].replace(",", ".")),
                        "nul": float(lignes[i + 3].replace(",", ".")),
                        "exterieur": float(lignes[i + 5].replace(",", ".")),
                    }
                except (ValueError, AttributeError):
                    break
                i += 6
            continue

        if titre == "Impair/Pair | Fin de Match":
            paires, i = _paires(lignes, i + 1, 2)
            if len(paires) == 2:
                cotes["pair_impair"] = {"pair": paires.get("Pair"), "impair": paires.get("Impair")}
            continue

        m_cages = re.match(r"^Cages Inviolées \| (.+) \| Fin de Match$", titre)
        if m_cages:
            nom = m_cages.group(1)
            paires, i = _paires(lignes, i + 1, 2)
            if len(paires) == 2:
                if nom == nom_domicile:
                    cotes["cages_inviolees_domicile"] = {"oui": paires.get("Oui"), "non": paires.get("Non")}
                elif nom == nom_exterieur:
                    cotes["cages_inviolees_exterieur"] = {"oui": paires.get("Oui"), "non": paires.get("Non")}
            continue

        if titre == "Score Exact | Fin de Match":
            paires, i = _paires_jusqua_rupture(lignes, i + 1, r"^(\d+-\d+|Other|Autre)$")
            scores = {k: v for k, v in paires.items() if re.match(r"^\d+-\d+$", k)}
            if scores:
                cotes["score_exact"] = scores
            continue

        if titre == "Nombre Exact de Buts | Fin de Match":
            paires, i = _paires_jusqua_rupture(lignes, i + 1, r"^(\d+|\d+\+)$")
            if paires:
                cotes["nombre_exact_buts"] = paires
            continue

        i += 1

    return cotes


if __name__ == "__main__":
    if len(sys.argv) != 4:
        print("Usage : python3 parse_betpawa.py fichier_brut.txt \"Nom Domicile\" \"Nom Exterieur\"", file=sys.stderr)
        sys.exit(1)
    with open(sys.argv[1], "r", encoding="utf-8") as f:
        texte = f.read()
    print(json.dumps(parse_betpawa(texte, sys.argv[2], sys.argv[3]), indent=2, ensure_ascii=False))
